### PR TITLE
Added support for dynamic layout renderer in log level filters (e.g. minLevel, maxLevel) 

### LINKED ISSUE #1502: Support layout renderer in log level filters (e.g. minLevel, maxLevel)
- Feature request

**NLog version**:  4.3.3)

**Platform**: .Net 4.0 , Mono4

I'd really love to be able to use layout renderers for my log level. Specifically my use case is - I need to be able to specify a default log level, but have this be overridable with an environment variable.

I tried the following but unfortunately I did not see any log output at all :(

``` xml
<rules>
  <logger name="*" minLevel="${whenEmpty:whenEmpty=TRACE:inner=${environment:variable=ST_LOGLEVEL}}" appendTo="console, syslog" />
</rules>
```

and this:

``` xml
<rules>
      <logger name="*" minLevel="${environment:variable=ST_LOGLEVEL:whenEmpty=TRACE}" appendTo="console, syslog" />
    </rules>
```

and finally this:

``` xml
<rules>
      <logger name="*" minLevel="${literal:text=TRACE}" appendTo="console, syslog" />
    </rules>
```

Which led me to believe that layout renderers are not available at all for log level.

### PR DESCRIPTION
**Pull request by snakefoot**
Extending #2709 and resolves #1502

allowes:

```xml
<nlog>
   <variable name='myLevel' value='debug'/>
    <rules>
      <logger minLevel='${myLevel}'/>
    </rules>
</nlog>
```

And you do this:

```c#
LogManager.KeepVariablesOnReload = true;
LogManager.Configuration.Variables["myLevel"] = "Info";
LogManager.Configuration.Reload();
```

### CODE REVIEW

**Comment by 304NotModified** [comment]
cool!

File: src/NLog/Config/LoggingRuleLevelFilter.cs [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+
+namespace NLog.Config
+{
+    using System;
+    using System.Collections.Generic;
+    using NLog.Common;
+    using NLog.Internal;
+    using NLog.Layouts;
+
+    internal interface ILoggingRuleLevelFilter
```
Comment (304NotModified) [inline src/NLog/Config/LoggingRuleLevelFilter.cs]: please move all interfaces and classes to a new file, thanks!

File: src/NLog/Config/LoggingRuleLevelFilter.cs [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+        public ILoggingRuleLevelFilter SetLoggingLevels(LogLevel minLevel, LogLevel maxLevel, bool enable)
+        {
+            if (!ReferenceEquals(LogLevels, Off.LogLevels))
+            {
+                for (int i = minLevel.Ordinal; i <= maxLevel.Ordinal; ++i)
+                    LogLevels[i] = enable;
+                return this;
+            }
+
+            return new LoggingRuleLevelFilter().SetLoggingLevels(minLevel, maxLevel, enable);
```
Comment (304NotModified) [inline src/NLog/Config/LoggingRuleLevelFilter.cs]: this is a bit unexpected. Why the `new LoggingRuleLevelFilter()`. The method is now a combination of "set" and "create"

File: src/NLog/Config/LoggingRuleLevelFilter.cs [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+            _levelFilter = levelFilter;
+            _activeFilter = new KeyValuePair<string, bool[]>(string.Empty, LoggingRuleLevelFilter.Off.LogLevels);
+        }
+
+        public ILoggingRuleLevelFilter SetLoggingLevels(LogLevel minLevel, LogLevel maxLevel, bool enable)
+        {
+            return new LoggingRuleLevelFilter(LogLevels).SetLoggingLevels(minLevel, maxLevel, enable);
+        }
+
+        private bool[] GenerateLogLevels()
```
Comment (304NotModified) [inline src/NLog/Config/LoggingRuleLevelFilter.cs]: please split this method in smaller ones, maybe some code could be deduplicated?

File: src/NLog/Config/LoggingRuleLevelFilter.cs [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+                return LoggingRuleLevelFilter.Off.LogLevels;
+
+            var activeFilter = _activeFilter;
+            if (!activeFilter.Key.Equals(new MinMaxLevels(minLevelFilter, maxLevelFilter)))
+            {
+                LogLevel minLevel = ParseLogLevel(minLevelFilter, LogLevel.MinLevel) ?? LogLevel.MaxLevel;
+                LogLevel maxLevel = ParseLogLevel(maxLevelFilter, LogLevel.MaxLevel) ?? LogLevel.MinLevel;
+
+                bool[] logLevels = new bool[LogLevel.MaxLevel.Ordinal + 1];
+                for (int i = minLevel.Ordinal; i <= Math.Min(logLevels.Length - 1, maxLevel.Ordinal); ++i)
```
Comment (304NotModified) [inline src/NLog/Config/LoggingRuleLevelFilter.cs]: PS I always use `i<= logLevels.Length - 1 && i < maxLevel.Ordinal` for those cases

File: src/NLog/Config/LoggingRuleLevelFilter.cs [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+                {
+                    logLevels[i] = true;
+                }
+                _activeFilter = activeFilter = new KeyValuePair<MinMaxLevels, bool[]>(new MinMaxLevels(minLevelFilter, maxLevelFilter), logLevels);
+            }
+            return activeFilter.Value;
+        }
+
+        LogLevel ParseLogLevel(string logLevel, LogLevel defaultLevel)
```
Comment (304NotModified) [inline src/NLog/Config/LoggingRuleLevelFilter.cs]: as there is an argument defaultLevel here, I think it would be logical to have also levelOnError here. Maybe rename defaultLevel to levelIfEmpty.

(defaults are sometimes also used for errors)

File: src/NLog/Config/LoggingRuleLevelFilter.cs (hunk 1 above)
Comment (snakefoot) [reply src/NLog/Config/LoggingRuleLevelFilter.cs]: I just repeated how it was done for `LoggerNameMatcher`. Have I done something different?

File: src/NLog/Config/LoggingRuleLevelFilter.cs (hunk 2 above)
Comment (snakefoot) [reply src/NLog/Config/LoggingRuleLevelFilter.cs]: Because it is difficult to combine dynamic filters with hardcoded assignment of levels. So it converts dynamic-filter to simple-filter if needed. If already simple-filter then it just modifies and returns this.

File: src/NLog/Config/LoggingConfigurationParser.cs [hunk 6]
```diff
@@ -669,31 +667,60 @@ private LoggingRule ParseRuleElement(ILoggingConfigurationElement loggerElement)
...
+                if (enableLevels.IndexOf('{') >= 0)
                 {
-                    rule.EnableLoggingForLevel(level);
+                    SimpleLayout simpleLayout = new SimpleLayout(enableLevels, _configurationItemFactory);
+                    simpleLayout.Initialize(this);
+                    rule.EnableLoggingForLevels(simpleLayout);
+                }
+                else
+                {
+                    if (enableLevels.IndexOf(',') >= 0)
```
[the remark this answers is not in the document]
Comment (snakefoot) [reply src/NLog/Config/LoggingConfigurationParser.cs]: Fixed

File: src/NLog/Config/LoggingRuleLevelFilter.cs (hunk 3 above)
Comment (snakefoot) [reply src/NLog/Config/LoggingRuleLevelFilter.cs]: Fixed

File: src/NLog/Config/LoggingRuleLevelFilter.cs (hunk 4 above)
Comment (snakefoot) [reply src/NLog/Config/LoggingRuleLevelFilter.cs]: Fixed

File: src/NLog/Config/LoggingRuleLevelFilter.cs (hunk 5 above)
Comment (snakefoot) [reply src/NLog/Config/LoggingRuleLevelFilter.cs]: Fixed

File: src/NLog/Config/LoggingRuleLevelFilter.cs (hunk 2 above)
Comment (snakefoot) [reply src/NLog/Config/LoggingRuleLevelFilter.cs]: Have now changed it into `GetSimpleFilterForUpdate()` that returns `LoggingRuleLevelFilter` (that has the method `SetLoggingLevels`)

File: src/NLog/Config/LoggingRuleLevelFilter.cs (hunk 1 above)
Comment (304NotModified) [reply src/NLog/Config/LoggingRuleLevelFilter.cs]: Yes, I think it was also a comment there. :/

For easier finding, refactoring and consistency, separate files are really preferred. 

If you move them when they are getting to big/complex, you will "lose" history in git.

Comment (snakefoot) [reply src/NLog/Config/LoggingRuleLevelFilter.cs]: Done

**Comment by snakefoot** [comment]
@304NotModified Not sure what to do now. Have made heavy refactoring of the unit-tests to make them simpler. And now you have committed.

**Comment by 304NotModified** [comment]
rebase or force push ;)

If you think it's better and rebase is too/unneeded difficult, then force push (and thus remove my commits).

**Comment by snakefoot** [comment]
@304NotModified Performed a manual rebase of all your relevant changes, and then a force push.

File: tests/NLog.UnitTests/Config/RuleConfigurationTests.cs [hunk 7]
```diff
@@ -589,5 +589,185 @@ public void LoggingRule_LevelOff_NotSetAsActualLogLevel()
...
+            expectedLogLevel = LogLevel.Fatal;
+            LogManager.Configuration.Variables["var_level"] = expectedLogLevel.ToString();
+            LogManager.ReconfigExistingLoggers();
+
+            for (int i = LogLevel.MinLevel.Ordinal; i <= LogLevel.MaxLevel.Ordinal; ++i)
+            {
+                if (LogLevel.FromOrdinal(i) == expectedLogLevel)
+                    Assert.True(b.IsEnabled(LogLevel.FromOrdinal(i)));
+                else
+                    Assert.False(b.IsEnabled(LogLevel.FromOrdinal(i)));
```
[the remark this answers is not in the document]
Comment (snakefoot) [reply tests/NLog.UnitTests/Config/RuleConfigurationTests.cs]: Fixed

File: tests/NLog.UnitTests/Config/RuleConfigurationTests.cs [hunk 8]
```diff
@@ -589,5 +591,221 @@ public void LoggingRule_LevelOff_NotSetAsActualLogLevel()
...
+            for (int i = LogLevel.MinLevel.Ordinal; i <= LogLevel.MaxLevel.Ordinal; ++i)
+            {
+                if (LogLevel.Fatal.Ordinal == i)
+                    Assert.True(logger.IsEnabled(LogLevel.FromOrdinal(i)));
+                else
+                    Assert.False(logger.IsEnabled(LogLevel.FromOrdinal(i)));
+            }
+        }
+
+        public static IEnumerable<object[]> LoggingRule_MinMaxLayout_ParseLevel_TestCases()
```
Comment (304NotModified) [inline tests/NLog.UnitTests/Config/RuleConfigurationTests.cs]: The idea by the [MemberData] was to reduce code duplication and make it easier to maintain. It seems we lost the latter one...

File: src/NLog/Config/LoggingRuleLevelFilter/DynamicLogLevelFilter.cs [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+            _levelFilter = levelFilter;
+            _activeFilter = new KeyValuePair<string, bool[]>(string.Empty, LoggingRuleLevelFilter.Off.LogLevels);
+        }
+
+        public LoggingRuleLevelFilter GetSimpleFilterForUpdate()
+        {
+            return new LoggingRuleLevelFilter(LogLevels);
+        }
+
+        private bool[] GenerateLogLevels()
```
Comment (304NotModified) [inline src/NLog/Config/LoggingRuleLevelFilter/DynamicLogLevelFilter.cs]: how critical is the performance of this method?

Comment (snakefoot) [reply src/NLog/Config/LoggingRuleLevelFilter/DynamicLogLevelFilter.cs]: What are you guiding towards?

File: tests/NLog.UnitTests/Config/RuleConfigurationTests.cs (hunk 8 above)
Comment (snakefoot) [reply tests/NLog.UnitTests/Config/RuleConfigurationTests.cs]: You are welcome to unroll all the loops and generate one giant list.

File: src/NLog/Config/LoggingRuleLevelFilter/DynamicLogLevelFilter.cs (hunk 9 above)
Comment (304NotModified) [reply src/NLog/Config/LoggingRuleLevelFilter/DynamicLogLevelFilter.cs]: refactoring

Comment (304NotModified) [reply src/NLog/Config/LoggingRuleLevelFilter/DynamicLogLevelFilter.cs]: by e.g replacing bool[] with a specialized class

Comment (snakefoot) [reply src/NLog/Config/LoggingRuleLevelFilter/DynamicLogLevelFilter.cs]: I can make refactoring for you without anyone being hurt.

File: tests/NLog.UnitTests/Config/RuleConfigurationTests.cs (hunk 8 above)
Comment (304NotModified) [reply tests/NLog.UnitTests/Config/RuleConfigurationTests.cs]: Is there a commit which I could use? Otherwise this will take some time.

File: src/NLog/Config/LoggingRuleLevelFilter/DynamicLogLevelFilter.cs (hunk 9 above)
Comment (snakefoot) [reply src/NLog/Config/LoggingRuleLevelFilter/DynamicLogLevelFilter.cs]: Refactoring complete

File: tests/NLog.UnitTests/Config/RuleConfigurationTests.cs (hunk 8 above)
Comment (snakefoot) [reply tests/NLog.UnitTests/Config/RuleConfigurationTests.cs]: Sorry don't have a commit ready in my pocket :)

Comment (304NotModified) [reply tests/NLog.UnitTests/Config/RuleConfigurationTests.cs]: I see these are now 80 test cases. I think this is hard to maintain and unneeded.  If we introduce a bug, it would be nice if 1 test fails, and not multiple. 

As rule of thumb I try to create these test cases:

- happy path
- alternatives paths 
- corner cases
- cases if needed for coverage.

That means no combination of cases, if it isn't needed for the list above.

For example this test: (shorter test, easier to show)

```c#
      [Theory]
        [InlineData("Off")]
        [InlineData("")]
        [InlineData((string)null)]
        [InlineData("Trace")]
        [InlineData("Debug")]
        [InlineData("Info")]
        [InlineData("Warn")]
        [InlineData("Error")]
        [InlineData(" error")]
        [InlineData("Fatal")]
        [InlineData("Wrong")]
        public void LoggingRule_LevelLayout_ParseLevel(string levelVariable
```
I think the cases "Debug", "Info", "Warn", "Error" (not " error") and "Fatal" not needed (redundant)

Comment (snakefoot) [reply tests/NLog.UnitTests/Config/RuleConfigurationTests.cs]: You are right about not needing all the official Levels. I just wanted to make absolutely sure I didn't have an issue with min/max-levels crossing.

File: src/NLog/Config/LoggingRuleLevelFilter/DynamicLogLevelFilter.cs (hunk 9 above)
Comment (304NotModified) [reply src/NLog/Config/LoggingRuleLevelFilter/DynamicLogLevelFilter.cs]: If I would create a class that encapsulates bool[], do you think that would be a bad thing for the performance?

Comment (snakefoot) [reply src/NLog/Config/LoggingRuleLevelFilter/DynamicLogLevelFilter.cs]: No hidding the bool[] array away and making methods instead on `ILoggingRuleLevelFilter` should not hurt performance.

Comment (snakefoot) [reply src/NLog/Config/LoggingRuleLevelFilter/DynamicLogLevelFilter.cs]: But please make that in a new PR instead of adding more into this one.

Comment (304NotModified) [reply src/NLog/Config/LoggingRuleLevelFilter/DynamicLogLevelFilter.cs]: Ok, before of after merge this PR?

Comment (snakefoot) [reply src/NLog/Config/LoggingRuleLevelFilter/DynamicLogLevelFilter.cs]: After merge is easier for me :)

**Comment by snakefoot** [comment]
@304NotModified Is the merge of this PR blocked by something?

**Comment by 304NotModified** [comment]
Same as the other PR, my free time at least.

Checked it in short, there are some things I like to get improved:

- LoggingRule_MinMaxLayout_ParseLevel_TestCases is really hard to maintain or even to reason which test cases this will generate. (maybe I have to check in VS2017, not done yet)
- There is some duplication in the testcases. As far as I remember, I tried to remove it, but it's still there/back?
- The ` LogLevels = new bool[LogLevel.MaxLevel.Ordinal + 1];` looks fragile and I prefer to encapsulate that in another object. I tried to help with some PRs, but have to conclude that commits of my on your branch aren't really helping - so I stopped with that for now.

**Comment by 304NotModified** [comment]
> Same as the other PR, my free time at least.
> 
> Checked it in short, there are some things I like to get improved:
> 
> * LoggingRule_MinMaxLayout_ParseLevel_TestCases is really hard to maintain or even to reason which test cases this will generate. (maybe I have to check in VS2017, not done yet)
> * There is some duplication in the testcases. As far as I remember, I tried to remove it, but it's still there/back?
> * The ` LogLevels = new bool[LogLevel.MaxLevel.Ordinal + 1];` looks fragile and I prefer to encapsulate that in another object. I tried to help with some PRs, but have to conclude that commits of my on your branch aren't really helping - so I stopped with that for now.

What's your opinion in this? This is still todo, isn't?

**Comment by snakefoot** [comment]
> @304NotModified  What's your opinion in this? This is still todo, isn't?

Sounded like refactorings that that you wanted to implement later. And since not part of the public interface, then it is not blocking the merge of this feature.

**Comment by 304NotModified** [comment]
> Sounded like refactorings that that you wanted to implement later. And since not part of the public interface, then it is not blocking the merge of this feature.

Sorry, It's is blocking for me. Too much tech debt in this one.

**Comment by snakefoot** [comment]
> Sorry, It's is blocking for me. Too much tech debt in this one.

So if I remove the LogLevel-array and replace with a class that has a property for each LogLevel, then you will be happy?

**Comment by 304NotModified** [comment]
i'm happy if the test cases aren't dynamic created and preferable with keeping things simple and reduce code duplication (I know that the latter two are a trade off)

**Comment by snakefoot** [comment]
> i'm happy if the test cases aren't dynamic created and preferable with keeping things simple and reduce code duplication

So you will be happy if I just reduce the test to a few success-tests, and remove all test-combination? (Leaving LogLevel-array as it is).

**Comment by 304NotModified** [comment]
yes, there is no need to test every combination (it's even considered bad style)

**Comment by snakefoot** [comment]
> yes, there is no need to test every combination (it's even considered bad style)

I just found this particular part of NLog very critical, and wanted to ensure it never broke. But will remove the yield-logic and just test a few good samples.

**Comment by snakefoot** [comment]
@304NotModified Have now simplified the input-generation for the unit-tests.

Also changed target-branch to master, so it can be included in NLog 4.6.7

**Comment by 304NotModified** [comment]
What's changed in the force push? (Can't find the difference)

**Comment by snakefoot** [comment]
> What's changed in the force push? (Can't find the difference)

No change beside pull + rebase,. Github was complaining it could not be merged right away.

**Comment by 304NotModified** [comment]
Ah , ok. You could ignore that warning. 

 I enabled that so it's clear before the merge it's targets the master ;)

**Comment by 304NotModified** [comment]
Merged :)

**Comment by 304NotModified** [comment]
Thanks!

**Comment by snakefoot** [comment]
Tried to update the wiki: https://github.com/NLog/NLog/wiki/Filtering-log-messages
